cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java

```diff
@@ -0,0 +1,226 @@
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Schema for OAuth store.
+ */
+public class OAuthStore {
+  private static final String OAUTH_PROVIDER_COL = "oauthprovider";
+  private static final String LOGIN_URL_COL = "loginurl";
+  private static final String TOKEN_REFRESH_URL_COL = "tokenrefreshurl";
+  private static final String CLIENT_CREDS_KEY_PREFIX = "oauthclientcreds";
+  private static final String REFRESH_TOKEN_KEY_PREFIX = "oauthrefreshtoken";
+  private static final Gson GSON = new Gson();
+  private static final Map<String, String> EMPTY_MAP = new HashMap<>();
+  private final TransactionRunner transactionRunner;
+  private final SecureStore secureStore;
+  private final SecureStoreManager secureStoreManager;
+
+  public static final StructuredTableId TABLE_ID = new StructuredTableId("oauth");
+  public static final StructuredTableSpecification TABLE_SPEC = new StructuredTableSpecification.Builder()
+      .withId(TABLE_ID)
+      .withFields(Fields.stringType(OAUTH_PROVIDER_COL),
+                  Fields.stringType(LOGIN_URL_COL),
+                  Fields.stringType(TOKEN_REFRESH_URL_COL))
+      .withPrimaryKeys(OAUTH_PROVIDER_COL)
+      .build();
+
+  public OAuthStore(
+      TransactionRunner transactionRunner,
+      SecureStore secureStore,
+      SecureStoreManager secureStoreManager) {
+    this.transactionRunner = transactionRunner;
+    this.secureStore = secureStore;
+    this.secureStoreManager = secureStoreManager;
+  }
+
+  /**
+   * Create/update an OAuth provider.
+   *
+   * @param oauthProvider {@link OAuthProvider} to write
+   * @throws OAuthStoreException if the write fails
+   */
+  public void writeProvider(OAuthProvider oauthProvider) throws OAuthStoreException {
+    try {
+      TransactionRunners.run(transactionRunner, context -> {
+        StructuredTable table = context.getTable(TABLE_ID);
+        table.upsert(getRow(oauthProvider));
+
+      }, TableNotFoundException.class, InvalidFieldException.class);
+    } catch (TableNotFoundException e) {
+      throw new OAuthStoreException("OAuth provider table not found", e);
+    } catch (InvalidFieldException e) {
+      throw new OAuthStoreException("OAuth provider object fields do not match table", e);
+    }
+
+    String namespace = NamespaceId.DEFAULT.getNamespace();
+    try {
+      secureStoreManager.put(
+          namespace,
+          getClientCredsKey(oauthProvider.getName()),
+          GSON.toJson(oauthProvider.getClientCredentials()),
+          "OAuth client creds",
+          EMPTY_MAP);
+    } catch (IOException e) {
+      throw new OAuthStoreException("Failed to write to OAuth provider secure storage", e);
+    } catch (Exception e) {
+      throw new OAuthStoreException("Namespace \"" + namespace + "\" does not exist", e);
```

Comment:
If it is namespace not found, shouldn't the exception being caught should be something like a NotFoundException?

Reply:
I would think so, but according to the Javadocs it just throws generic Exception when the namespace isn't found: [link](https://github.com/cdapio/cdap/blob/83034a85e61137e0c0c4984e0f98c31529c04bf6/cdap-api/src/main/java/io/cdap/cdap/api/security/store/SecureStoreManager.java#L40).